### ISSUE TITLE
[BUG] Concurrent write to _delta_log on AWS S3 Delta Table results in data loss

### ISSUE THREAD

**Issue by ArtOfStew** [delta-io/delta#1830]
### Describe the problem

We have the following stack: Delta 2.2.0 running in AWS Glue 4.0 jobs with PySpark.

**TL;DR:** Multiple parallel glue jobs writing to the same table can concurrently write to the same _delta_log transaction log file. The later job "wins"; it performs an overwrite on the S3 object, rather than creating a new one.

**Longer version:**

We are using the above Glue jobs to load data from flat files (on S3) into Delta Tables. We often have multiple files to load to the same table as part of the same batch; in this case we spawn several Glue Jobs at once, aiming to run them in parallel. The files may be written to different partitions or not; this issue still manifests.

We've found that data intermittently goes missing when we process multiple files. We have traced this to a concurrent write to the transaction log.

We have read up on articles on Delta Table concurrency; most of them are around read concurrency. This bug appears to be linked to S3 interactions in particular.

The result of this is that data partitions exist, but the transaction log has no record of them. This, in turn, means that generating the symlink manifest does not have these partitions, which means our downstream process either don't have the data (athena) or break (redshift spectrum). Basically, it results is the table breaking ACIDity.

We have a workaround we have implemented; I will describe that below. However, it would be great if this could be resolved in the framework (if possible). If there is a better way of parallelizing loads or a setting to ensure we don't overwrite log files, please let us know.

#### Steps to reproduce

I cannot supply the full stacks of code we use, but I will describe our setup in more detail. I also have a number of screen shots demonstrating our analysis of the situation.

1. Set up a glue job with PySpark. The script should load a file from S3 to a delta table. We do not read from the existing delta table before insertion. The Spark job simply transforms the file columns into a dataframe and writes to the table; we set up the tables beforehand via configuration.

Here is the spark session we set up:
```
def get_spark_session():
    builder = SparkSession.builder.appName("MyApp") \
        .config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension") \
        .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog")\
        .config("spark.delta.logStore.class","io.delta.storage.S3SingleDriverLogStore")

    spark = configure_spark_with_delta_pip(builder).getOrCreate()

    return spark
```

Here is the eventual write statement. The partition columns are as shown, but the table has arbitrary other columns. The replace condition is actually generated dynamically, of course.
```
replace_cond = "edl_extract_dt='2023-01-01 01:01:01' and edl_file_name='test_col_basic_overwrite_source_20230101010101.csv'"
partition_columns = ['edl_extract_dt','edl_file_name']
data_df.write.format('delta') \
      .mode('overwrite') \
      .partitionBy(partition_columns) \
      .option("replaceWhere", replace_cond) \
      .save(destination_s3_uri)
```

2. Have a large number of files to write in the S3 source location. I've demonstrated this consistently with 20 files.
3. Run the glue job once for each file, simultaneously. We use Step Functions to orchestrate this.
4. Generate the symlink manifest for the Delta table after all of the above runs are complete (this just makes things easier, you could also just look at the transaction log.
5. Examine the partitions written to the Delta location vs the partitions according to the _symlink_format_manifest folder. You *should* find a discrepancy.

#### Observed results

Here are the contents of the delta table location after the above steps. There are 20 partitions, from the 20 files.
![Table contents on s3](https://github.com/delta-io/delta/assets/32175288/fd68beb0-0125-48cb-976e-3a03321b85a3)

Here are the _symlink_format_manifest contents: several partitions are missing.
![symlink format manifest](https://github.com/delta-io/delta/assets/32175288/0d8f578a-9760-4ab9-8b57-b68f02041ef2)

Here is a view of the version history of the contents of the _delta_log folder. You can see some files have multiple overwrites, such as version ...007.json
![delta log versions](https://github.com/delta-io/delta/assets/32175288/57982a46-cfa5-456f-87f3-cc4566eca2c8)

Earliest to latest for version ...007.json:
- Partition ...03.csv was written first.
- 03 was then overwritten by 16
- 16 was then overwritten by 13. From above, 13 is the only partition present in the symlink manifest folder.
![transaction log 007 versions](https://github.com/delta-io/delta/assets/32175288/91502d58-3add-400a-9357-76cb02e39325)

#### Expected results

Expectation: transaction log should be consistent with all data writes that occur.

#### Further details: Workaround

We have implemented a kind of semaphore: Before we write to a delta table, we use a DynamoDB Table to ensure only one write happens at a time. This slows things down a bit, of course, but it's not too significant in our case.

### Environment information

* Delta Lake version: 2.2.0
* Spark version: 3.3 (Glue 4.0)
* Scala version: 2 (Glue 4.0)

### Willingness to contribute

The Delta Lake Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the Delta Lake code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the Delta Lake community.
- [x] No. I cannot contribute a bug fix at this time.

**Comment by ArtOfStew** [comment]
@nkarpov Makes sense, thanks for the link... We didn't see it before implementing our workaround. Pity, but the workaround looks like it's on version 2.4.0 onwards, and we're restricted to 2.2.0 at least for now, so I'm not too broken up about doing our own workaround. And, I guess we were on the right track, since we're using Dynamo to do essentially the same thing.

I'll close off the "bug"!

**Comment by tdas** [comment]
The DynamoDB solution has exists since Delta 1.2 - https://delta.io/blog/2022-05-18-multi-cluster-writes-to-delta-lake-storage-in-s3/

So you should be able to very easily use it in Delta 2.2.

**Comment by ArtOfStew** [comment]
Okay great - thanks for the additional resource, @tdas. We'll give that configuration a shot on our end. 👍

**Comment by watfordkcf** [comment]
@tdas we had to use a custom authentication provider in order to support it on our setup: https://github.com/watfordkcf/aws-custom-credential-provider/tree/com.kcftech.sd/delta-spark-utils

**Comment by tdas** [comment]
Whoa! Can you explain why did you have to build a custom authentication provider? Why didnt the usual default credential provider chain work? 

cc @scottsand-db

**Comment by watfordkcf** [comment]
We needed to assume a specific role on EMR Serverless. The default AWS credential providers were inadequate for this (not for a lack of trying). It seems Zillow had the same problem many moons ago and it was never really resolved.

**Comment by petersonjr** [comment]
Hi everybody,
I was wondering, since I saw the news that [AWS now provides S3 conditional writes](https://aws.amazon.com/pt/about-aws/whats-new/2024/08/amazon-s3-conditional-writes/), whether Delta Lake would support multi-cluster concurrent writes in S3 without the need for DynamoDB.
I just wanted to check if this is on the roadmap. Thank you!

@tdas

**Comment by tandriamil** [comment]
The [Delta Rust connector already seems to have implemented it by default](https://github.com/delta-io/delta-rs/pull/3091) for Amazon S3, MinIO and R2. However, I am not aware of such implementation of the other connectors.

Can we get an update or a detailed table on the state of this support for each connector ? Thank you !

**Comment by foss-contributor** [comment]
@ArtOfStew, thank you for your contribution! This issue has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this issue or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!

**Comment by ArtOfStew** [comment]
@foss-contributor Thanks for the prompt - yes, have been using Delta Lake again recently, and it's come a long way - this has been resolved over the last 18 months. I will close.
